vms/platformvm/warp/README.md

```diff
@@ -0,0 +1,121 @@
+```
+
+## Sending an Avalanche Warp Message
+
+A Blockchain on Avalanche sends an Avalanche Warp Message by coming to agreement on the message that every validator should be willing to sign. As an example, the VM of a Blockchain may define that once a block is accepted, the VM should be willing to sign a message including the block hash in the payload to attest to any other Subnet that the block was accepted. The contents of the payload, how to aggregate the signature (VM-to-VM communication, off-chain relayer, etc.), is left to the VM.
+
+Once the validator set of a blockchain is willing to sign the message M, an aggregator performs the following process:
+
+1. Gather signatures of the message M from N validators (where the N validators meet the required threshold of stake on the destination chain)
+2. Aggregate the N signatures into a multi-signature
+3. Look up the canonical validator set at the P-Chain height where the message will be verified
+4. Encode the selection of the N validators included in the signature in a bitset
+5. Construct the Signed Message from the aggregate signature, bitset, and unsigned message
+
+## Verifying / Receiving an Avalanche Warp Message
+
+Avalanache Warp Messages are verified within the context of a specific P-Chain height included in the [ProposerVM](../../proposervm/README.md)'s header. The P-Chain height is provided as context to the wrapped VM when verifying the VM's internal blocks (implemented by the optional interface [WithVerifyContext](../../../snow/engine/snowman/block/block_context_vm.go)).
+
+To verify the message, the wrapped VM utilizes this Warp package to perform the following steps:
+
+1. Lookup the canonical validator set of the Subnet sending the message at the P-Chain height
+2. Filter the canonical validator set to only the validators claimed by the signature
+3. Verify the weight of the included validators meets the required threshold (threshold defined by the receiving VM)
+4. Aggregate the public keys of the claimed validators into a single aggregate public key
+5. Verify the aggregate signature of the unsigned message against the aggregate public key
+
+Once a message is verified, it is left to the VM to define the semantics of delivering a verified message.
+
+## Design Considerations
+
+### Processing Historical Avalanche Warp Messages
+
+Verifying an Avalanche Warp Message requires a lookup of validator sets at a specific P-Chain height. The P-Chain serves lookups by maintaining reverse diffs that can be iterated in order to re-create the validator set of any Subnet at a specific height.
+
+As a chain moves forward, the number of reverse diffs that need to be computed in order to look up the validator set needed to verify an Avalanche Warp Messages increases over time.
+
+Therefore, in order to support verifying historical Avalanche Warp Messages, VMs should provide a mechanism to determine whether an Avalanche Warp Message was treated as valid or invalid within a historical block.
```

Comment:
```suggestion
Therefore, in order to support verifying historical Avalanche Warp Messages, VMs should provide a mechanism to determine whether an Avalanche Warp Message was treated as valid or invalid within a historical block range.
```

Reply:
Why block range? Warp messages are verified within a single block as opposed to a range. Is this because we're verifying a range of historical blocks?

Going to leave this out because I think it's clearer as is.